// angular-disable-browser-back-button/fesm5/angular-disable-browser-back-button.js
import { __decorate, __param } from 'tslib';
import { Injectable, Inject, NgModule } from '@angular/core';
import { NavigationEnd, Router } from '@angular/router';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/router';
function _window() {
    return window;
}
var WindowRef = /** @class */ (function () {
    function WindowRef() {
    }
    Object.defineProperty(WindowRef.prototype, "nativeWindow", {
        get: function () {
            return _window();
        },
        enumerable: true,
        configurable: true
    });
WindowRef.ɵfac = function WindowRef_Factory(t) { return new (t || WindowRef)(); };
WindowRef.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: WindowRef, factory: function (t) { return WindowRef.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(WindowRef, [{
        type: Injectable
    }], function () { return []; }, null); })();
    return WindowRef;
}());
var BackButtonDisableModule = /** @class */ (function () {
    function BackButtonDisableModule(router, windowRef, preserveScrollPosition) {
        this.router = router;
        this.windowRef = windowRef;
        this.preserveScrollPosition = preserveScrollPosition;
        this.scrollX = 0;
        this.scrollY = 0;
        this.window = this.windowRef.nativeWindow;
        this.disableBackButton();
        this.addPopStateEventListener();
    }
    BackButtonDisableModule_1 = BackButtonDisableModule;
    BackButtonDisableModule.forRoot = function (config) {
        return {
            ngModule: BackButtonDisableModule_1,
            providers: [
                {
                    provide: 'preserveScrollPosition',
                    useValue: config && 'preserveScrollPosition' in config ? config.preserveScrollPosition : false
                }
            ]
        };
    };
    BackButtonDisableModule.prototype.addPopStateEventListener = function () {
        var _this = this;
        this.window.addEventListener('popstate', function () {
            if (_this.preserveScrollPosition)
                _this.getScrollCoordinates();
            _this.window.history.pushState(null, null, null);
            if (_this.preserveScrollPosition)
                setTimeout(_this.scrollToThePreviousPosition.bind(_this));
        });
    };
    BackButtonDisableModule.prototype.scrollToThePreviousPosition = function () {
        this.window.scrollTo(this.scrollX, this.scrollY);
    };
    BackButtonDisableModule.prototype.getScrollCoordinates = function () {
        this.scrollX = this.window.scrollX;
        this.scrollY = this.window.scrollY;
    };
    BackButtonDisableModule.prototype.disableBackButton = function () {
        var _this = this;
        this.window.history.pushState(null, null, null);
        this.router.events.subscribe(function (event) {
            if (event instanceof NavigationEnd) {
                _this.window.history.pushState(null, null, null);
            }
            ;
        });
    };
    var BackButtonDisableModule_1;
    BackButtonDisableModule.ctorParameters = function () { return [
        { type: Router },
        { type: WindowRef },
        { type: Boolean, decorators: [{ type: Inject, args: ['preserveScrollPosition',] }] }
    ]; };
    BackButtonDisableModule = BackButtonDisableModule_1 = __decorate([ __param(2, Inject('preserveScrollPosition'))
    ], BackButtonDisableModule);
BackButtonDisableModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: BackButtonDisableModule });
BackButtonDisableModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function BackButtonDisableModule_Factory(t) { return new (t || BackButtonDisableModule)(ɵngcc0.ɵɵinject(ɵngcc1.Router), ɵngcc0.ɵɵinject(WindowRef), ɵngcc0.ɵɵinject('preserveScrollPosition')); }, providers: [WindowRef] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BackButtonDisableModule, [{
        type: NgModule,
        args: [{
                providers: [WindowRef]
            }]
    }], function () { return [{ type: ɵngcc1.Router }, { type: WindowRef }, { type: Boolean, decorators: [{
                type: Inject,
                args: ['preserveScrollPosition']
            }] }]; }, null); })();
    return BackButtonDisableModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { BackButtonDisableModule, WindowRef };
